plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionData.java

```diff
@@ -150,7 +153,12 @@ public static Object getValue(JsonNode partitionValue, Type type)
                     throw new UncheckedIOException("Failed during JSON conversion of " + partitionValue, e);
                 }
             case DECIMAL:
-                return partitionValue.decimalValue();
+                Types.DecimalType decimalType = (Types.DecimalType) type;
+                return Decimals.rescale(
+                        partitionValue.decimalValue(),
+                        DecimalType.createDecimalType(
+                                decimalType.precision(),
+                                decimalType.scale()));
```

Comment:
the method here has undocumented semantics. Should it return Trino stack represaentation? should it return some intermediate representation?

- if it is supposed to return Trino stack representation, then `FLOAT: return partitionValue.floatValue()` and `STRING: partitionValue.asText()` cases are incorrect too (first should do `floatToIntBits` and the second should wrap into `Slice`)
- it it is supposed to return SomethingElse™, then the scaling handling may not belong here and may belong somewhere else

in any case, we should document our findings what the method is supposedly doing

Reply:
This is part of the serialization and deserialization of the partition spec to/from JSON https://iceberg.apache.org/spec/#appendix-c-json-serialization.

So we are representing this in the Iceberg metadata spec format. The issue is with our implementation of this serialization and deserialization.
There are no tests on this class as well. It should round-trip properly (current issue I fixed was that a `BigDecimal("12.30")` didn't round-trip and instead became `BigDecimal("12.3")` on read.

I'll extract the decimal changes to a new branch to follow-up if we want to address the related issues together.